### ISSUE TITLE
Building large responses is slow

### ISSUE THREAD

**Issue by coding-chimp** [rmosolgo/graphql-ruby#3410]
**Describe the bug**

We have a query where clients can load the shape of a zip code to draw it onto a map. The DB call is pretty simple and for the most part, this query performs pretty well, however, there a few handfuls of zip codes in Germany, which shapes are so complicated that we're generating a pretty big response. For the worst offender, we end up generating 401kb of JSON. Building those large responses is really slow with GraphQL.

Just to put some actual numbers on it, for a "simple" zip code (8kb response) the request takes ~20ms on my machine while it takes ~680ms for the "complicated" zip code (401kb response). On production, it actually takes multiple seconds, but that feels like there is something else going on that we need to debug.

For now, we just slapped some caching on top of the query and we're fine. The data doesn't ever really change anyway. But it feels like this shouldn't be this slow.

I simplified and extracted the query including the GeoJSON of one zip code into a [gist](https://gist.github.com/coding-chimp/0f32f0c33fafd41c122a5f97f72754be). I've already done some profiling with v.1.12.6 and #3364. The results are also in the gist.

This is probably related to #2352, #3345, and #3379.

**Comment by coding-chimp** [comment]
Actually, when I benchmark the extracted code, the performance deteriorates pretty much linearly compared to the initial GeoJSON size. A 50x bigger GeoJSON means it'll take ~50x longer to build the response. So I guess there is nothing inherently wrong here.

**Comment by rmosolgo** [comment]
Have you considered only supporting square-shaped jurisdictions 😆 

> The data doesn't ever really change anyway. But it feels like this shouldn't be this slow.

That's the problem  -- GraphQL-Ruby can't tell the difference between complicated, important fields and trivial, unimportant fields, so it adds the same overhead to all those calls (tracing, `.dup`'ed paths to support non-linear execution, bookkeeping for promises and other lazy values). 

Thanks for providing the detailed instrumentation output.

It looks like you're using `GraphQL::Dataloader` but _not_ `GraphQL::Batch`, is that true? If so, one possibility would be trying to extract "lazy value" handling from the runtime, that would remove these lines from the profile: 

```
        32   (9.4%)          32   (9.4%)     Concurrent::Collection::MriMapBackend#compute_if_absent
        28   (8.2%)          28   (8.2%)     GraphQL::Schema.lazy_methods
         7   (2.1%)           7   (2.1%)     Concurrent::Collection::NonConcurrentMapBackend#[]
```

The other big component there is the type-at-path tracking: 

```
        40  (11.7%)          38  (11.1%)     GraphQL::Execution::Interpreter::Runtime#set_non_null_at
        10   (2.9%)          10   (2.9%)     GraphQL::Execution::Interpreter::Runtime#continue_value
         9   (2.6%)           9   (2.6%)     GraphQL::Execution::Interpreter::Runtime#dead_path?
```

IIRC, that's used for: 

- Propagating `nil` when an non-null field returns an invalid `null` (we have to know where in the response we can write `nil` 
- In non-linear execution, knowing when _not_ to write a value to the response because its parent object was removed because of null propagation 

Maybe some more efficient approach could be implemented, but ... I don't know _what_ 😖

**Comment by swalkinshaw** [comment]
Would using the raw value feature help here since it short-circuits execution? It's kind of an escape hatch since I think it bypasses guarantees about the response being correct (according to schema)? But you could do some manual validation first which would be optimized in a single place.

**Comment by rmosolgo** [comment]
That's a great workaround suggestion!

**Comment by swalkinshaw** [comment]
I don't think it's documented, but here's the PR @coding-chimp which explains it: https://github.com/rmosolgo/graphql-ruby/pull/2699

**Comment by coding-chimp** [comment]
Raw value looks perfect for this use case. I did a quick test and it is about 20x faster. Thanks, @swalkinshaw!

> It looks like you're using GraphQL::Dataloader but not GraphQL::Batch, is that true? If so, one possibility would be trying to extract "lazy value" handling from the runtime

Yes, that's true and I think that's a great idea! But how would that work? I guess lazy execution could be made an optional feature in v2, but I'm not sure what could be done until then? 🤔

**Comment by coding-chimp** [comment]
I'm gonna close this since my immediate problem was solved by the raw value feature.

I plan to investigate extracting the lazy execution from the runtime, but I'll need to take some time to wrap my head around it first. So I might open another issue or PR for that at some point.

**Comment by rmosolgo** [comment]
Just thought I'd follow up: I just released 1.12.11 which included a big perf improvement to the same lines pointed out in the trace above (#3494). Please try that new version if you're still having any trouble with this!
